### ISSUE TITLE
Xcode 14 introduces new run script build phase warning for SwiftLint

### ISSUE THREAD

**Issue by opfeffer** [realm/SwiftLint#4015]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

Possibly somewhat of a duple of #2783, but Xcode 14 appears to emit a new warning for run script build phases without specified outputs:
> warning build: Run script build phase 'SwiftLint' will be run during every build because it does not specify any outputs. To address this warning, either add output dependencies to the script phase, or configure it to run in every build by unchecking "Based on dependency analysis" in the script phase.

<img width="616" alt="image" src="https://user-images.githubusercontent.com/1138127/177389768-d33f6dd7-7a5a-4d26-8c8f-4585fc9b072b.png">

I'm following the usage guidance as documented in the [README.md](https://github.com/realm/SwiftLint/#xcode). I'm curious what the proper configuration should be for to appease this warning as by default SwiftLint doesn't really modify any files. That said, it seems excessive to run in every build if no input files have changed....

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
export PATH="$PATH:/opt/homebrew/bin"

if which swiftlint > /dev/null; then
  swiftlint
else
  echo "warning: SwiftLint not installed, download from https://github.com/realm/SwiftLint"
fi

```

### Environment

- Installed via Homebrew
   ```
    $ swiftlint version
    0.47.1
    ```
- default swiftlint config
- Xcode 14.0 beta 2 (14A5229c)

**Comment by samitaha2** [comment]
Also following! Thank you for opening up this task, I was just about to do the same

**Comment by kamaal111** [comment]
Under the script you can uncheck the checkbox with the label `Based on dependency analysis` and this warning will go away

**Comment by opfeffer** [comment]
> Under the script you can uncheck the checkbox with the label `Based on dependency analysis` and this warning will go away

You're right that checking that box makes this warning disappear. But afaict that doesn't really solve the bigger issue that swiftlint runs on _every_ file regardless of modification status during incremental builds. Depending on the size of your codebase this can have significant impact on compile time.

Would love to find a way to use Xcode's input/output files to only run swiftlint on the files that actually changed maybe? 🙏

**Comment by cdhoffmann** [comment]
> > Under the script you can uncheck the checkbox with the label `Based on dependency analysis` and this warning will go away
> 
> You're right that checking that box makes this warning disappear. But afaict that doesn't really solve the bigger issue that swiftlint runs on _every_ file regardless of modification status during incremental builds. Depending on the size of your codebase this can have significant impact on compile time.
> 
> Would love to find a way to use Xcode's input/output files to only run swiftlint on the files that actually changed maybe? 🙏

100%

**Comment by yosi199** [comment]
> > Under the script you can uncheck the checkbox with the label `Based on dependency analysis` and this warning will go away
> 
> You're right that checking that box makes this warning disappear. But afaict that doesn't really solve the bigger issue that swiftlint runs on _every_ file regardless of modification status during incremental builds. Depending on the size of your codebase this can have significant impact on compile time.
> 
> Would love to find a way to use Xcode's input/output files to only run swiftlint on the files that actually changed maybe? 🙏

+1 on that, looking for a solution

**Comment by fahad-drivary** [comment]
> > Under the script you can uncheck the checkbox with the label `Based on dependency analysis` and this warning will go away
> 
> You're right that checking that box makes this warning disappear. But afaict that doesn't really solve the bigger issue that swiftlint runs on _every_ file regardless of modification status during incremental builds. Depending on the size of your codebase this can have significant impact on compile time.
> 
> Would love to find a way to use Xcode's input/output files to only run swiftlint on the files that actually changed maybe? 🙏

100% agree

**Comment by AndreaMiotto** [comment]
Any news on this? I can see some PRs have been merged but I'm still having the warning. And im not sure if disabling the Xcode checkmark is a good idea.

**Comment by jifang** [comment]
This is what we do for running swiftlint only on modified files (based on git status)
```
git status --porcelain | grep -v '^ \?D' | cut -c 4- | sed 's/.* -> //' | tr -d '"' | while read file 
do
    if [[ "$file" == *.swift ]]; then
        swiftlint "$file" --fix
    fi
done
```

**Comment by jifang** [comment]
> > Under the script you can uncheck the checkbox with the label `Based on dependency analysis` and this warning will go away
> 
> You're right that checking that box makes this warning disappear. But afaict that doesn't really solve the bigger issue that swiftlint runs on _every_ file regardless of modification status during incremental builds. Depending on the size of your codebase this can have significant impact on compile time.
> 
> Would love to find a way to use Xcode's input/output files to only run swiftlint on the files that actually changed maybe? 🙏
This is what we do for running swiftlint only on modified files (based on git status)
```
git status --porcelain | grep -v '^ \?D' | cut -c 4- | sed 's/.* -> //' | tr -d '"' | while read file 
do
    if [[ "$file" == *.swift ]]; then
        swiftlint "$file" --fix
    fi
done
```

**Comment by mywristbands** [comment]
> This is what we do for running swiftlint only on modified files (based on git status)
> 
> ```
> git status --porcelain | grep -v '^ \?D' | cut -c 4- | sed 's/.* -> //' | tr -d '"' | while read file 
> do
>     if [[ "$file" == *.swift ]]; then
>         swiftlint "$file" --fix
>     fi
> done
> ```

Thanks for that, but I still want to make sure _all_ files are passing the linter. What I'd like is for Xcode to only run swiftlint on files that have been changed since the last incremental build.

**Comment by KrisLau** [comment]
https://github.com/realm/realm-swift/issues/7891#issuecomment-1192447849

**Comment by Gobans** [comment]
If anyone who using Tuist, modify TargetAction option make sure `basedOnDependencyAnalysis` would be false.
``` swift
import ProjectDescription

public extension TargetScript {
    static let SwiftLintString = TargetScript.pre(script: """
if test -d "/opt/homebrew/bin/"; then
    PATH="/opt/homebrew/bin/:${PATH}"
fi

export PATH="$PATH:/opt/homebrew/bin"

if which swiftlint > /dev/null; then
    swiftlint
else
    echo "warning: SwiftLint not installed, download from https://github.com/realm/SwiftLint"
fi
""", name: "SwiftLintString", basedOnDependencyAnalysis: false) // here change the option
}

```
Because of this warning, I was suffered a little 😂 
Wish this comment could be helpful

**Comment by kenji21** [comment]
> > This is what we do for running swiftlint only on modified files (based on git status)
> > ```
> > git status --porcelain | grep -v '^ \?D' | cut -c 4- | sed 's/.* -> //' | tr -d '"' | while read file 
> > do
> >     if [[ "$file" == *.swift ]]; then
> >         swiftlint "$file" --fix
> >     fi
> > done
> > ```
> 
> Thanks for that, but I still want to make sure _all_ files are passing the linter. What I'd like is for Xcode to only run swiftlint on files that have been changed since the last incremental build.

we even do use the "Product -> Analyse" to lint all files (and disable for previews/IBDesignables ones) :
```
if [[ "$BUILD_DIR" == *"IBDesignables"* ]] || [[ "$BUILD_DIR" == *"Previews"* ]] ; then
    echo "not linting for IBDesignables/SwiftUI Previews builds";
    exit 0
fi

if [[ $RUN_CLANG_STATIC_ANALYZER == "YES" ]] ; then
    time swiftlint
else 
[.....]
    time swiftlint --use-script-input-files
fi
```

**Comment by razor313** [comment]
None of the previous solutions worked for me. do you have any suggestion, preferably a simple one? :)))

**Comment by alexanderkhitev** [comment]
Hey guys! Do we have any news/updates? Thanks

**Comment by kaiserabliz** [comment]
Still watching this thread.

**Comment by steven851007** [comment]
I've solved the issue by adding a pre-build action script that dynamically generates the input files before every build for SwitLint. This setup correctly solves the warning and also improves the incremental build times. You can find a sample project with this setup here, with the script included and with an explanation of how it works:

https://github.com/steven851007/SwiftLint_build_phase_example

**Comment by kaiserabliz** [comment]
> I've solved the issue by adding a pre-build action script that dynamically generates the input files before every build for SwitLint. This setup correctly solves the warning and also improves the incremental build times. You can find a sample project with this setup here, with the script included and with an explanation of how it works:
> 
> https://github.com/steven851007/SwiftLint_build_phase_example

@steven851007  Thanks, this is really interesting, Looking forward to testing it out today.

**Comment by cabyambo** [comment]
Has there been any official fix for this?

**Comment by jonnyijapan** [comment]
Threads like [this one on SO](https://stackoverflow.com/questions/74012563/warning-build-run-script-build-phase-module-will-be-run-during-every-build-be) could be worth a follow.

**Comment by omar3alaa** [comment]
Ok, first of all,  As per [this](https://developer.apple.com/documentation/Xcode/improving-the-speed-of-incremental-builds#Declare-inputs-and-outputs-for-custom-scripts-and-build-rules) you need to specify two things in order for the script doesn't run with every incremental build:

1. Input files (Files that needed as inputs in the script, if those files are not changed, then the script won't run in this case)
2. Output files (Files that are used as outputs after running the script, if this files are missing, then the script will run every time)

 ## 1. Input files: ##

-  Here you can put all the files that need swiftlint to be run at, and that would be hassle to add every and each file manually.
- What you need to do here is put at least the main folder path, so in case you have project named `Project`, that has a folder `Sources` which contains all the project's swift files, so you need to specify this folder path as input file -> `$(SRCROOT)/Sources/`
- If you have more than module in the same xcworkspace so you need to specify all of those projects paths in the script with the alternative project's name -> `$(SRCROOT)/{module_name}/` (Replace `{module_name}` with your module's name without the `{}`)
- You'll have something like that eventually:
 
<img width="704" alt="Screen Shot 2023-03-15 at 5 39 00 PM" src="https://user-images.githubusercontent.com/30074316/225325758-26fd4329-bb0d-4aaa-9238-67d6a9461a60.png">


 ## 2. Output files: ##
- Since SwiftLint doesn't produce any generated files, so we can put a static file that could be used as output file just for the sake of the script doesn't run with every build without even changing any lines of code
- So first inside the script will put this: 
```
# This output is used by Xcode 'outputs' to avoid re-running this script phase.
echo "SUCCESS" > "${SCRIPT_OUTPUT_FILE_0}"
```
- Then we'll add an output file -> `$(DERIVED_FILE_DIR)/swiftlintResults.txt`

## 3. Final look: ##
- Finally, you'll have script phase looks like this below

<img width="843" alt="Screen Shot 2023-03-15 at 5 42 25 PM" src="https://user-images.githubusercontent.com/30074316/225326734-b26980c5-de32-4f77-a6ab-5d599e35ef70.png">


Try to build, first time for sure it will re-run the script because of the changes we just made in this script phase, after building successfully, try to build once again, you'll notice that the warning is not there anymore because we already added the input and output files needed, and the script won't run with every build until you change in the files that were put as input files. 👍

**Comment by Iikeli** [comment]
Has there been any work on this? Or is there a good hack that works?

**Comment by jrtibbetts** [comment]
I'm seriously considering removing the SwiftLint build phase entirely until this is fixed. I can run it manually as needed, and keep it in the CI/CD flow.

**Comment by mildm8nnered** [comment]
> I'm seriously considering removing the SwiftLint build phase entirely until this is fixed. I can run it manually as needed, and keep it in the CI/CD flow.

This is a feature of, and these warnings are coming from, Xcode, and there's not very much that SwiftLint can do about it. 

For any run script build phase, you will either need to uncheck the "Based on dependency analysis" checkbox, in which case you will see a message, but not a warning per se, about how the script will be run on every build, or leave the box checked, but not specify any output and input files, in which case you will see the warning, or you can add input and output files, as per the referenced SO thread, or the other suggestions above.

**Comment by jfahrenkrug** [comment]
I'm not familiar with the SwiftLint codebase, but here's a question: Would adding a flag like `--incremental` or something along those lines be an option? It could write two files into - maybe - a `.swiftlint` directory. One file that contains info about the files that have been linted in the last run, (maybe with `last modified` values or a hash), and a file that contains info about the linting results for the last run. If those two files are used as the output params in Xcode's run script build phase, and all swift files in the project are used as the input.... would that solve this issue? Or is that a naive idea and I'm missing something?

**Comment by mildm8nnered** [comment]
> I'm not familiar with the SwiftLint codebase, but here's a question: Would adding a flag like `--incremental` or something along those lines be an option? It could write two files into - maybe - a `.swiftlint` directory. One file that contains info about the files that have been linted in the last run, (maybe with `last modified` values or a hash), and a file that contains info about the linting results for the last run. If those two files are used as the output params in Xcode's run script build phase, and all swift files in the project are used as the input.... would that solve this issue? Or is that a naive idea and I'm missing something?

You could just `touch somefile` in your Run Script build phase, and make `somefile` your output file, and that would be pretty much exactly functionally equivalent to the above.

**Comment by rajivd15** [comment]
@opfeffer Did you find a good solution for this issue?
I do see similar warnings for Crashlytics, GraphQL build phases. One thing I tried was to just add an Output file to the build phase as I do not have any inputs to the script -`$(DERIVED_FILE_DIR)/Pods-checkSwiftLintResult.txt` and the warning didn't show up. 
I just wanted to see if someone can confirm - would all the SwiftLint errors will show up in that file? not sure how CI/CD pipeline will be affected by this change.

**Comment by nguyenlam96** [comment]
hi, any updates to optimize build time? thanks

**Comment by jonnyijapan** [comment]
Just an idea... remove swiftlint from Xcode build phase(s), and only use it before GIT commits?

**Comment by MakarUrbanov** [comment]
Guys just add `$(DERIVED_FILE_DIR)/swiftlint_output` into your swiftlint build phrase to output files list and the warning is gone

My xcodegen:
```
postCompileScripts:
      - path: "Scripts/swiftLint.sh"
        name: SwiftLint
        outputFiles: ["$(DERIVED_FILE_DIR)/swiftlint_output"]
```

**Comment by jonnyijapan** [comment]
> Guys just add `$(DERIVED_FILE_DIR)/swiftlint_output` into your swiftlint build phrase to output files list and the warning is gone
> 
> 
> 
> My xcodegen:
> 
> ```
> 
> postCompileScripts:
> 
>       - path: "Scripts/swiftLint.sh"
> 
>         name: SwiftLint
> 
>         outputFiles: ["$(DERIVED_FILE_DIR)/swiftlint_output"]
> 
> ```

I tried this but and yes it got rid of that message but actually it now fails to add comment on lines that failed lint testing. So I assume that the lint testing stops working with this change. Maybe it depends on what you have in that swiftLint.sh file?

**Comment by MakarUrbanov** [comment]
> > Guys just add `$(DERIVED_FILE_DIR)/swiftlint_output` into your swiftlint build phrase to output files list and the warning is gone
> > My xcodegen:
> > ```
> > 
> > postCompileScripts:
> > 
> >       - path: "Scripts/swiftLint.sh"
> > 
> >         name: SwiftLint
> > 
> >         outputFiles: ["$(DERIVED_FILE_DIR)/swiftlint_output"]
> > ```
> 
> I tried this but and yes it got rid of that message but actually it now fails to add comment on lines that failed lint testing. So I assume that the lint testing stops working with this change. Maybe it depends on what you have in that swiftLint.sh file?

Yeah you're right. My solution doesn't work 🙃
So now in this case, I don’t know how to get rid of this warning, but “Based on dependency analysis” solves the problem on small projects

**Comment by NikKovIos** [comment]
The task is not just remove a warning. Swiftlint takes a time to lint every time. So we need to note lint not changed files.

**Comment by litoarias** [comment]
> ```swift
> basedOnDependencyAnalysis
> ```

Maybe is so later to respond this for 'xcodegen' users, but for if anyme helps I use the following code to clean the warning:
```shell
....
postCompileScripts:
   - path: "Scripts/swiftLint.sh"
     name: SwiftLint
     basedOnDependencyAnalysis: false
....
```

**Comment by litoarias** [comment]
> > > Guys just add `$(DERIVED_FILE_DIR)/swiftlint_output` into your swiftlint build phrase to output files list and the warning is gone
> > > My xcodegen:
> > > ```
> > > 
> > > postCompileScripts:
> > > 
> > >       - path: "Scripts/swiftLint.sh"
> > > 
> > >         name: SwiftLint
> > > 
> > >         outputFiles: ["$(DERIVED_FILE_DIR)/swiftlint_output"]
> > > ```
> > 
> > 
> > I tried this but and yes it got rid of that message but actually it now fails to add comment on lines that failed lint testing. So I assume that the lint testing stops working with this change. Maybe it depends on what you have in that swiftLint.sh file?
> 
> Yeah you're right. My solution doesn't work correctly 🙃 So now in this case, I don’t know how to get rid of this warning, but “Based on dependency analysis” solves the problem on small projects

Maybe is so later to respond this for 'xcodegen' users, but for if anyme helps I use the following code to clean the warning:
```shell
....
postCompileScripts:
   - path: "Scripts/swiftLint.sh"
     name: SwiftLint
     basedOnDependencyAnalysis: false
....
```

**Comment by DmitryKhristenko** [comment]
Hi! 
Any updates here?

**Comment by NikKovIos** [comment]
Any updates?

**Comment by SimplyDanny** [comment]
Isn't SwiftLint's built-in cache not helping here? You uncheck "Based in dependency analysis" and always run SwiftLint on all files. But due to the cache, most results will not be recomputed.

**Comment by NikKovIos** [comment]
@SimplyDanny 

```
lint="${PODS_ROOT}/SwiftLint/swiftlint"
$lint --fix && $lint
```
takes 18 seconds

**Comment by SimplyDanny** [comment]
Is this without modified files? How many files are we talking about in total?

**Comment by nacho4d** [comment]
I have a mid size project (1~2 hundred of files).
This script will gather all changed files and pass ONCE to swift-format. This is faster than calling swift-format per file.

```sh
    changedPaths=$(git status -s . | awk '
    /^D/ {next}
    !/\.swift$/ {next}
    /^R|^C/ {print $4; next}
    {print $2}
    ' | xargs -n1)
    if [ -n "$changedPaths" ]; then
        echo "Lint and Format Swift Sources: $changedPaths"
        check_swift_format
        swift-format format --in-place --configuration "$CONFIG_FILE" $changedPaths
        swift-format lint --configuration "$CONFIG_FILE" $changedPaths
    else
        echo "Lint and Format Swift Sources: No Swift files have changed."
    fi
```

My approach uses awk because it can do grep/sed/cut/tail/etc... all by himself.

**Comment by NikKovIos** [comment]
https://github.com/steven851007/SwiftLint_build_phase_example Can it help?

**Comment by DataGreed** [comment]
Any solutions to this?

**Comment by rokas-ambrazevicius** [comment]
`
#!/bin/bash
set -euo pipefail

# Ensure Homebrew bin on Apple Silicon
if [[ "$(uname -m)" == "arm64" ]]; then
    export PATH="/opt/homebrew/bin:$PATH"
fi

# Go to repository root (ensures relative paths are correct)
REPO_ROOT=$(git rev-parse --show-toplevel)
cd "$REPO_ROOT"

if command -v swiftlint >/dev/null 2>&1; then
    echo "Running SwiftLint (no-cache) autocorrect for modified Swift files…"

    # Collect modified Swift files (excluding deletions)
    git status --porcelain | \
        grep -v '^ \?D' | \
        cut -c 4- | \
        sed 's/.* -> //' | \
        tr -d '"' | \
        while read -r file; do
            if [[ "$file" == *.swift && -f "$file" ]]; then
                echo "Linting & correcting: $file"
                
                # Run linting first (fresh, no cache if available)
                if swiftlint lint --help | grep -q -- "--no-cache"; then
                    swiftlint lint --no-cache "$file" || true
                else
                    swiftlint lint "$file" || true
                fi
                
                # Then run autocorrect (fresh, no cache if supported)
                if swiftlint autocorrect --help 2>&1 | grep -q -- "--no-cache"; then
                    swiftlint autocorrect --no-cache "$file" || true
                else
                    swiftlint autocorrect "$file" || true
                fi
            fi
        done

else
    echo "warning: SwiftLint not installed. Install via Homebrew:"
    echo "  brew install swiftlint"
fi
`
Trying this out, based on the above comments.
